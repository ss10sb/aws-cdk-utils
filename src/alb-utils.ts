import {Stack} from "@aws-cdk/core";
import {ApplicationLoadBalancer, IApplicationLoadBalancer} from "@aws-cdk/aws-elasticloadbalancingv2";
import {Config} from "./config";
import {Utils} from "./utils";
import {SsmUtils} from "./ssm-utils";

export class AlbUtils {

    public static getAlbByArn(stack: Stack, arn: string, prefix: string = 'stack'): IApplicationLoadBalancer {
        return ApplicationLoadBalancer.fromLookup(stack, `${prefix}-alb`, {
            loadBalancerArn: arn
        });
    }

    public static getAlbArnParamKey(config: Config, name: string = 'alb01'): string {
        return `${AlbUtils.getBaseAlbName(config, name)}-arn`;
    }

    public static getArnFromParams(stack: Stack, key: string): string {
        return SsmUtils.getValue(stack, key);
    }

    public static getDefaultAlbName(config: Config, name: string = 'alb01'): string {
        return `${this.getBaseAlbName(config, name)}/alb`;
    }

    public static getBaseAlbName(config: Config, name: string = 'alb01'): string {
        return `${Utils.getBaseName(config)}-${name}`;
    }
}