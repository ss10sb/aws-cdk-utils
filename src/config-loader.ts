import * as fs from 'fs';
import * as path from "path";
import {merge} from 'lodash';
import {Config} from "./config";

export class ConfigLoader<T extends Config> {
    readonly configDir: string;

    constructor(configDir: string) {
        this.configDir = configDir;
    }

    private async getConfigFromFiles(env: string | null): Promise<T> {
        const defaultEnv = await this.getFromBase('defaults');
        let overrideEnv = {};
        if (env) {
            overrideEnv = await this.getFromBase(env);
        }
        return <T>merge(defaultEnv, overrideEnv);
    }

    private async getFromBase(base: string): Promise<object> {
        const jsFile = path.resolve(this.configDir, `${base}.js`);
        if (fs.existsSync(jsFile)) {
            const results = await import(jsFile);
            return results.default ?? {};
        }
        const jsonFile = path.resolve(this.configDir, `${base}.json`);
        if (fs.existsSync(jsonFile)) {
            return JSON.parse(fs.readFileSync(jsonFile, 'utf8'));
        }
        console.log(`Environment '${base}' not found.`);
        return {};
    }

    public async load(env: string): Promise<T> {
        return await this.getConfigFromFiles(env);
    }

    public static convertStringToConfig<T extends Config>(value: string): T {
        try {
            return <T>JSON.parse(value);
        } catch (e) {
            console.log("Error parsing JSON", value);
        }
        return <T>{};
    }
}