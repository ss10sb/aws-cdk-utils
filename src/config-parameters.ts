export interface ConfigParameters {
    readonly vpcId?: string;
}
