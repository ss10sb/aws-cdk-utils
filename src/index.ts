export * from './alb-utils';
export * from './config';
export * from './config-loader';
export * from './config-mutable';
export * from './config-param-store';
export * from './config-parameters';
export * from './config-stack';
export * from './deep-writable';
export * from './newable';
export * from './ssm-utils';
export * from './utils';
export * from './vpc-utils';