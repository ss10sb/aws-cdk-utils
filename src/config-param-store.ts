import {Construct, Stack} from "@aws-cdk/core";
import {Config} from "./config";
import {SsmUtils} from "./ssm-utils";
import {IStringParameter, StringParameter} from "@aws-cdk/aws-ssm";
import {ConfigLoader} from "./config-loader";

export class ConfigParamStore extends Construct {

    public storeKeys: String[] = ['Name', 'College', 'Environment', 'StoreConfig', 'Parameters'];

    constructor(scope: Construct, id: string) {
        super(scope, id);
    }

    store<T extends Config>(name: string, config: T): StringParameter {
        let stored: { [key: string]: any } = {};
        for (const key of this.storeKeys) {
            stored[key.toString()] = config[key as keyof T];
        }
        const stringed: string = JSON.stringify(stored);
        return SsmUtils.createParam(this, this.getParamName(name), stringed);
    }

    fetchStringAsValue<T extends Config>(name: string): T {
        return ConfigLoader.convertStringToConfig<T>(this.fetchStringParameterAsString(name));
    }

    fetchStringParameterAsString(name: string): string {
        return SsmUtils.getStringValue(this, name);
    }

    fetchStringAsPlaceholder(name: string): IStringParameter {
        return SsmUtils.getStringParam(this, this.getParamName(name));
    }

    fetchSecretAsPlaceholder(name: string): IStringParameter {
        return SsmUtils.getSecretParam(this, this.getParamName(name));
    }

    getParamName(name: string): string {
        return `/${this.node.id}/${name}`;
    }

    getArn(name: string): string {
        return `arn:aws:ssm:${Stack.of(this).region}:${Stack.of(this).account}:parameter${this.getParamName(name)}`;
    }
}