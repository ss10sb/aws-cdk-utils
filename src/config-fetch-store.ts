import {Construct} from "@aws-cdk/core";
import {IStringParameter} from "@aws-cdk/aws-ssm";
import {Config} from "./config";
import {ConfigParamStore} from "./config-param-store";
import {deepMerge} from "aws-cdk/lib/util";
import {NonConstruct} from "./non-construct";

export class ConfigFetchStore<T extends Config> extends NonConstruct {
    readonly configParamStore: ConfigParamStore;
    public paramName: string = 'config';

    constructor(scope: Construct, id: string) {
        super(scope, id);
        this.configParamStore = new ConfigParamStore(scope, id);
    }

    fetch(): T {
        return <T>this.retrieveConfigValueFromParamStore();
    }

    store(config: T): IStringParameter {
        return this.configParamStore.store<T>(this.paramName, config);
    }

    getArn(): string {
        return this.configParamStore.getArn(this.paramName);
    }

    getName(): string {
        return this.configParamStore.getParamName(this.paramName);
    }

    protected storeConfigToParamStore(config: T): IStringParameter | null {
        return this.configParamStore.store<T>(this.paramName, config);
    }

    protected retrieveConfigValueFromParamStore(): T | null {
        try {
            return this.configParamStore.fetchStringAsValue<T>(this.paramName);
        } catch (e) {
            console.log('Unable to retrieve parameter', e);
        }
        return null;
    }
}