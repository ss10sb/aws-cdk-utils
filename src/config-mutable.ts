import {Construct} from "@aws-cdk/core";
import {IStringParameter} from "@aws-cdk/aws-ssm";
import {Config} from "./config";
import {ConfigParamStore} from "./config-param-store";
import {deepMerge} from "aws-cdk/lib/util";

export class ConfigMutable<T extends Config> {
    readonly scope: Construct;
    readonly id: string;
    readonly configParamStore: ConfigParamStore;
    public param: IStringParameter | null;
    public paramName: string = 'config';

    constructor(scope: Construct, id: string) {
        this.scope = scope;
        this.id = id;
        this.configParamStore = new ConfigParamStore(scope, id);
        this.param = null;
    }

    mutate(config: T): T {
        if (this.wantsToStoreConfig(config)) {
            if (!this.wantsInitialize(config)) {
                const paramConfig = this.retrieveConfigValueFromParamStore();
                if (paramConfig && !this.isEmpty(paramConfig)) {
                    config = paramConfig;
                }
            }
            this.param = this.storeConfigToParamStore(config);
        }
        return config;
    }

    getParamArn(): string {
        return this.configParamStore.getArn(this.paramName);
    }

    protected mergeConfigs(config: T, paramConfig: T | null): T {
        return <T>deepMerge(paramConfig || {}, config);
    }

    protected storeConfigToParamStore(config: T): IStringParameter | null {
        return this.configParamStore.store<T>(this.paramName, config);
    }

    protected retrieveConfigValueFromParamStore(): T | null {
        try {
            return this.configParamStore.fetchStringAsValue<T>(this.paramName);
        } catch (e) {
            console.log('Unable to retrieve parameter', e);
        }
        return null;
    }

    protected wantsToStoreConfig(config: T): boolean {
        return config?.StoreConfig ?? false;
    }

    protected wantsInitialize(config: T): boolean {
        return config?.Initialize ?? false;
    }

    protected isEmpty(obj: Object): boolean {
        return Object.keys(obj).length === 0;
    }
}