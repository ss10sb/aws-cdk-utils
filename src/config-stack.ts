import * as cdk from '@aws-cdk/core';
import {Config} from './config';
import {ConfigMutable} from "./config-mutable";

export class ConfigStack<T extends Config> extends cdk.Stack {

    public config: T;
    public readonly internalId: string;
    protected configMutable: ConfigMutable<T> | undefined;

    constructor(scope: cdk.Construct, id: string, stackProps: cdk.StackProps, config: T, suffix: string = '') {
        const internalId = id;
        if (suffix) {
            id = `${id}-${suffix.toLowerCase()}`;
        }
        super(scope, id, stackProps);
        this.internalId = internalId;
        this.config = this.mutateConfig(config);
        this.preInit();
        this.init();
        this.postInit();
    }

    get isProd(): boolean {
        return this.config.Environment === 'prod';
    }

    mixNameWithId(name: string): string {
        return `${this.internalId}-${name}`;
    }

    preInit(): void {
        // do pre init stuff here
    }

    init(): void {
        // do stuff here
    }

    postInit(): void {
        // do post init stuff here
    }

    protected mutateConfig(config: T): T {
        this.configMutable = new ConfigMutable<T>(this, this.mixNameWithId('config'));
        return this.configMutable.mutate(config);
    }
}
