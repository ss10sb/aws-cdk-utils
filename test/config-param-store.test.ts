import {App} from '@aws-cdk/core';
import {SynthUtils} from "@aws-cdk/assert";
import {Config, ConfigParamStore, ConfigStack} from "../src";

describe('config param store', () => {
    it('can store params', () => {
        const app = new App();
        const buildConfig = <Config>{
            Name: 'test'
        }
        const stack = new ConfigStack(app, 'test', {}, buildConfig);
        const configParamStore = new ConfigParamStore(stack, 'cps');
        const param = configParamStore.store('test', buildConfig);
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    });
    it('can store advanced params', () => {
        const app = new App();
        const buildConfig = <Config>{
            Name: 'a'.repeat(4100)
        }
        const stack = new ConfigStack(app, 'test', {}, buildConfig);
        const configParamStore = new ConfigParamStore(stack, 'cps');
        const param = configParamStore.store('test', buildConfig);
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    });
    it('can fetch string param as placeholder', () => {
        const app = new App();
        const buildConfig = <Config>{
            Name: 'test'
        }
        const stack = new ConfigStack(app, 'test', {}, buildConfig);
        const configParamStore = new ConfigParamStore(stack, 'cps');
        const param = configParamStore.fetchStringAsPlaceholder('test');
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    });
    it('can create arn', () => {
        const app = new App();
        const buildConfig = <Config>{
            College: 'PCC',
            Environment: 'sdlc',
            Name: 'test'
        }
        const stack = new ConfigStack(app, 'test', {}, buildConfig);
        const configParamStore = new ConfigParamStore(stack, stack.mixNameWithId('param'));
        const arn = configParamStore.getArn('defaults');
        expect(arn).toContain(':parameter/test-param/defaults');
    });
});
