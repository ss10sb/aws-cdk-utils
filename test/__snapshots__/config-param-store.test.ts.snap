// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config param store can fetch string param as placeholder 1`] = `
Object {
  "Parameters": Object {
    "cpstestlookupParameter": Object {
      "Default": "/cps/test",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
}
`;

exports[`config param store can store advanced params 1`] = `
Object {
  "Resources": Object {
    "testcpstestparam85CAD9FA": Object {
      "Properties": Object {
        "Name": "/cps/test",
        "Tier": "Advanced",
        "Type": "String",
        "Value": "{\\"Name\\":\\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\\"}",
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;

exports[`config param store can store params 1`] = `
Object {
  "Resources": Object {
    "testcpstestparam85CAD9FA": Object {
      "Properties": Object {
        "Name": "/cps/test",
        "Tier": "Standard",
        "Type": "String",
        "Value": "{\\"Name\\":\\"test\\"}",
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;
